gluten-data/src/main/scala/io/glutenproject/vectorized/ColumnarBatchSerializer.scala

```diff
@@ -124,106 +123,102 @@ private class ColumnarBatchSerializerInstance(
+      ShuffleReaderJniWrapper
+        .create()
+        .readStream(shuffleReaderHandle, byteIn),
+      nmm)
 
-      override def readKey[T: ClassTag](): T = {
-        // We skipped serialization of the key in writeKey(), so just return a dummy value since
-        // this is going to be discarded anyways.
-        null.asInstanceOf[T]
-      }
+    private var cb: ColumnarBatch = _
 
-      @throws(classOf[EOFException])
-      override def readValue[T: ClassTag](): T = {
-        if (cb != null) {
-          cb.close()
-          cb = null
-        }
-        val batch = {
-          val maybeBatch =
-            try {
-              wrappedOut.next()
-            } catch {
-              case ioe: IOException =>
-                this.close()
-                logError("Failed to load next RecordBatch", ioe)
-                throw ioe
-            }
-          if (maybeBatch == null) {
-            // EOF reached
-            this.close()
-            throw new EOFException
+    private var numBatchesTotal: Long = _
+    private var numRowsTotal: Long = _
+
+    private var isClosed: Boolean = false
+
+    override def asIterator: Iterator[Any] = {
+      // This method is never called by shuffle code.
+      throw new UnsupportedOperationException
+    }
+
+    override def readKey[T: ClassTag](): T = {
+      // We skipped serialization of the key in writeKey(), so just return a dummy value since
+      // this is going to be discarded anyways.
+      null.asInstanceOf[T]
+    }
+
+    @throws(classOf[EOFException])
+    override def readValue[T: ClassTag](): T = {
+      if (cb != null) {
+        cb.close()
+        cb = null
+      }
+      val batch = {
+        val maybeBatch =
+          try {
+            wrappedOut.next()
+          } catch {
+            case ioe: IOException =>
+              this.close()
+              logError("Failed to load next RecordBatch", ioe)
+              throw ioe
           }
-          maybeBatch
+        if (maybeBatch == null) {
+          // EOF reached
+          this.close()
+          throw new EOFException
         }
-        val numRows = batch.numRows()
-        logDebug(s"Read ColumnarBatch of $numRows rows")
-        numBatchesTotal += 1
-        numRowsTotal += numRows
-        cb = batch
-        cb.asInstanceOf[T]
+        maybeBatch
       }
+      val numRows = batch.numRows()
+      logDebug(s"Read ColumnarBatch of $numRows rows")
+      numBatchesTotal += 1
+      numRowsTotal += numRows
+      cb = batch
+      cb.asInstanceOf[T]
+    }
 
-      override def readObject[T: ClassTag](): T = {
-        // This method is never called by shuffle code.
-        throw new UnsupportedOperationException
-      }
+    override def readObject[T: ClassTag](): T = {
+      // This method is never called by shuffle code.
+      throw new UnsupportedOperationException
+    }
 
-      override def close(): Unit = {
-        if (!isClosed) {
-          TaskResources.removeResource(resourceId)
```

Comment:
If remove this call, when will `releaseResource` be called?

Reply:
I'm still confused. Shouldn't somewhere explicitly call `releaseResource`? If nowhere calls `releaseResource`, when will the `TaskResource` be removed from the registry?

Reply:
@marin-ma We release all corresponding resources when the Spark task is getting stopped. https://github.com/oap-project/gluten/blob/3fe4c959e07739ac142433485691f92a8a75bfd1/gluten-core/src/main/scala/org/apache/spark/util/TaskResources.scala#L171

By adding `releaseResource(id: String)` we can releasing a single registered resource by ID.

So after then `releaseAll` could be considered batch version of `releaseResource`.

Reply:
The idea of https://github.com/oap-project/gluten/pull/4276 is to remove the single resource immediately when `close` is called, to avoid caching lots of object references. We shouldn't rely on `releaseAll` to release the object at task end, because with lots of objects being cached, GC will take a very long time.

Reply:
Thanks for the explanation. And IIUC This PR doesn't alter the core logic you mentioned above. 

See `remove` will be finally called through `releaseResource`:

https://github.com/oap-project/gluten/pull/4292/files#diff-8f21fe60b00b9ff709fad96d28ddb7a06c7f5e20ae7929f9216844de663e636aR103-R105

https://github.com/oap-project/gluten/pull/4292/files#diff-8f21fe60b00b9ff709fad96d28ddb7a06c7f5e20ae7929f9216844de663e636aR232-R233

Reply:
@marin-ma , I double-check the PR and noticed it was my mistake not adding `releaseResource` into deserializer code. It should be fixed in latest commit.

Thanks for catching!